drivers/console/bm_serial.c

```diff
@@ -61,105 +57,293 @@ K_SEM_DEFINE(decode_sem, 0, 1);
+    -1, -1, -1, -1, -1, 0xB, 0xA, -1, -1, 0x9, 0x8, -1, -1, -1, -1, -1,
+    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+    -1, -1, -1, -1, -1, 0x7, 0x6, -1, -1, 0x5, 0x4, -1, -1, -1, -1, -1,
+    -1, -1, -1, -1, -1, 0x3, 0x2, -1, -1, 0x1, 0x0, -1, -1, -1, -1, -1,
+    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+};
+
+#define USER0_NODE DT_ALIAS(user0)
+#define USER1_NODE DT_ALIAS(user1)
+
+/* These map to D3/D4 on the L496ZG Nucleo*/
+static const struct gpio_dt_spec user0 = GPIO_DT_SPEC_GET(USER0_NODE, gpios);
+static const struct gpio_dt_spec user1 = GPIO_DT_SPEC_GET(USER1_NODE, gpios);
+
+static bm_parse_ret_t bm_serial_align(uint8_t *rx_byte)
+{
+    bm_parse_ret_t ret = {.new_state=BM_ALIGN, .success=0};
+	static uint8_t preamble_ctr = 0;
 
-    // Awaiting complete frame
-    if ( await_alignment )
+	if (*rx_byte == BM_PREAMBLE_VAL) 
 	{
-		// If incoming byte is 0, we can now align to the next frame
-		if ( byte_in == 0 )
-		{
-			await_alignment = 0;
-			return;
-		}
+		preamble_ctr++;
+	}
+	else if (preamble_ctr >= BM_PREAMBLE_LEN && *rx_byte == BM_DELIMITER_VAL) 
+	{
+		/* We have aligned with the Preamble and Delimiter*/
+		preamble_ctr = 0;
+		ret.new_state = BM_COLLECT_HEADER;
+		ret.success = 1;
 	}
 	else
 	{
-		// Collect bytes until either the next delimiter is reached, or we overflow our len
-		if( byte_in == 0 )
-		{
-			// Frame complete
-			//LOG_INF( "Len=%d", len );
+		/* We reach this case if we receive non-preamble bytes when we expect them */
+		LOG_ERR("Expected Preamble, but got errant bytes");
+
+		/* Reset counter */
+		preamble_ctr = 0;
+	}
+    return ret;
+}
+
+bm_ret_t bm_serial_process_byte(uint8_t byte)
+{
+	bm_ret_t test_ret = {.retval=-EINPROGRESS, .length=0, .buf_ptr=NULL};
+	bm_parse_ret_t ret;
+	static uint16_t payload_ctr = 0;
+	static bm_parse_state_t _state = BM_ALIGN;
+	static uint8_t decoded_version = 0;
+	static uint8_t decoded_type = 0;
+	static uint16_t decoded_length = 0;
+	static uint16_t decoded_crc = 0;
+	int8_t decoded_nibble_lsb = 0;
+	int8_t decoded_nibble_msb = 0;
+	uint16_t computed_crc16 = 0;
+	bm_frame_header_t rx_header;
+
+	switch (_state)
+	{
+		case BM_ALIGN:
+		
+			/* Reset static variables */
+			decoded_length = 0;
+			decoded_version = 0;
+			decoded_type = 0;
+			decoded_crc = 0;
+			payload_ctr = 0;
+
+			ret = bm_serial_align(&byte);
+			_state = ret.new_state;
+			break;
+		case BM_COLLECT_HEADER:
+			encoded_rx_buf[write_buf_idx].buf[encoded_rx_ctr++] = byte;
 
-			if (len != 0)
+			/* Collect until we have a full frame header */
+			if (encoded_rx_ctr == (2*sizeof(bm_frame_header_t)))
```

Comment:
Since this if statement decides if the rest of the code is run or not, you can simplify it a bit and reduce "depth" by changing it to 
```
if (encoded_rx_ctr < (2*sizeof(bm_frame_header_t))) {
    break;
}
```

Then move the rest of the code which had been inside the if statement to the left. :D

Reply:
Unrelated, but I see we're waiting unit we have enough bytes for a whole header before checking. If we get one extra byte before the header, the check will fail. Do we discard the whole thing, or discard the first byte, then check again with the rest of the data?

Reply:
I think in this case we are going to be discarding the frame and waiting until we see the preamble/delimiter pattern again.